### DISCUSSION TITLE
Do you use port forwarding to test local builds? - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by lukecookssw** [SSWConsulting/SSW.Rules.Content#5203] [Chewing the Fat]
Contributors: @danielmackay

Let's talk about this rule:
[Do you use port forwarding to test local builds?](https://www.ssw.com.au/rules/port-forwarding/)

**Comment by GordonBeeming** [comment]
I didn't realize when I first saw the video that dev tunnels had been released yet, I'd used it with dev containers and github code spaces but not through visual studio yet.

I've used ngrok for many years with a custom domain which is great for demos which is where I've mostly used it.

**Comment by SamBWagner** [comment]
I think this Rule's awesome! I do think that tightly tying the rule to Visual Studio is not great but alas, a Rider user can dream haha.

**Comment by jernejk** [comment]
There are so many dev tools for port forwarding, there was a time, where we could do it all for free.
Nowadays, a lot of tools are either no longer updated to work with modern tools or they locked many features behind paywall, not all of them updated to work properly as they used to.

I think VS port forwarding is very timely and necessary today.

**Comment by drwharris** [comment]
Best tool out there for Port forwarding if you are using VS.

**Comment by jakebayliss** [comment]
Great feature added to VS. I'll try pointing a Dynamics env to a dev tunnel and see how we go!

**Comment by bahjatalaadelssw** [comment]
Any tool that speeds up the feedback and review process is ok with me.

**Comment by ozdemir-mehmet** [comment]
cool feature

**Comment by AlexRothwell** [comment]
Knowing about the port forwarding feature in VS will be incredibly valuable, especially for giving a PO quick access to an application before it has been deployed

**Reply by AlexRothwell** [reply]
I've just showed this feature to the scrum master/project manager for a client and they love it. I can see it being very useful for showcases too

**Comment by matt-goldman** [comment]
DevTunnels are awesome. One more issue for them to fix for me which is host header forwarding, than I'll scrap ngrok and PacketRiot altogether. There are some other cool features in the pipeline too!

**Comment by tkapa** [comment]
This looks pretty cool 👍

**Comment by JackDevAU** [comment]
Looks super quick and easy to setup and use! Can't wait to try it out. Awesome stuff on the video and rule 🚀

**Comment by piers-sinclair** [comment]
I had heard about dev tunnels from Daniel, but I didn't draw the connection to port forwarding until this CTF!

I've been using port forwarding for dev containers in VS Code and it's pretty important for the process.

**Comment by camillars** [comment]
Dev Tunnels seems like a useful tool for developers looking to streamline their testing and feedback processes. :)

**Comment by cvionnet** [comment]
Very nice feature, and looks very easy to configure and run.

ℹ️ A presentation video published a few days ago on dotnet YT channel: https://www.youtube.com/watch?v=B9K9eseNcKE

**Comment by tombui99** [comment]
Dev Tunnels seems to be a much better alternative compared to ngrok, I'd definitely use it for future client if needed.

**Comment by UlyssesMaclaren** [comment]
ok so my understanding is that the most common use case of dev tunnels is to enable remote debugging of web applications running on a local machine.... very cool for our distrobuted teams :)

**Comment by zacharykeeping** [comment]
Seems like a great feature for VS. Excited to use it.

**Comment by AntPolkanov** [comment]
Cool tool, thank you!

One little thing was missed I think. In what edition of Visual Studio Dev Tunnels is available? All of them or only Professional/Enterprise?
Can we please add this info to the rule?

**Reply by danielmackay** [reply]
@AntPolkanov - good suggestion.  I'm running VS Enterprise.  Would need to look into the other version.  Thanks!

**Comment by JeanThirion** [comment]
This sounds pretty cool ! 
I've never used it, but am likely to give it a go next time 🌠

**Comment by amankumarrr** [comment]
Cool feature! It saves time without interrupting the whole staging environment, especially if you're working with a large team. If the feature you're working on has an impact on the entire project and you're unsure whether it will be included in the release or not, you can test it before deploying it to staging, which is a great advantage.

**Comment by Ben0189** [comment]
Awesome tools in visual studio, I will try to use that next time

**Comment by danielmackay** [comment]
Was really excited to discover this tool in VS.  Made debugging a Web API sitting behind APIM very easy.  Enjoy all! 😄

**Comment by vladislav-kir** [comment]
I have previously used ngrok to test webhooks when integrating with 3d party software. It was really straightforward to use it, unlike another non-mentioned alternative - nginx. 
Dev Tunnels sound even better, because they are integrated inside Visual Studio!

**Comment by steven0x51** [comment]
Dev Tunnels is a very handy port forwarding tool when using VS

**Comment by leehawkinsssw** [comment]
I was not familiar with this feature and it looks like a handy way of opening up the testing of a locally-running application across the internet. This should be very helpful for POs and other client stakeholders to experience the actual application before it's deployed to production.

**Comment by bradystroud** [comment]
I use port forwarding for mobile dev everyday, its cool that this is integrated into VS now 🙂

I made a draft PR to improve the rule, let me know what you think
https://github.com/SSWConsulting/SSW.Rules.Content/pull/5210

**Comment by wicksipedia** [comment]
Looks like a nice feature for people who live in VS

**Comment by chrisschultzssw** [comment]
Looks like a really good solution if you use Visual Studio - I wonder if other IDEs will release similar features to make port forwarding easier.

**Comment by christoment** [comment]
It's a good solution for people with VS. Would definitely use if it is available outside of VS :)

**Comment by landonmaxwell** [comment]
Nice, my favourite part is the butterfly in the video 👌

**Comment by AshAnil** [comment]
It's a good solution for people using Visual Studio.

**Comment by Harry-Ross** [comment]
This seems like a really great tool to use with Visual Studio. However, I would prefer to use command line based options, as it means I don't have to be tied to Visual Studio.

**Comment by Geordie88** [comment]
while I don't have much to add to the underlying VS debate that seems to be running through this discussion, I do appreciate tools that allow for robust testing and I like how the rule gives a step by step tutourial

**Comment by jimmy1992abc** [comment]
Not a dev, but the video from Dan clearly explained the concept to me. I think this is a handy tool to test the work done. Meanwhile, I am curious if there could be a tool to test any developers' work without pre-set the environment (something like the tool itself providing all the framework).

**Comment by Levijj22** [comment]
Sounds like a great tool, very well explained Daniel!

**Comment by KahaMason** [comment]
This sounds like a very interesting approach to the test / review / feedback process. If we can streamline the process with this, that's awesome!

**Comment by Abf09** [comment]
Interesting to know how to use port forwarding for local test builds, I'll give this a try for my next project since I use Vs  Code.

**Comment by ozair-ashfaque** [comment]
Definitely use it in future

**Comment by stephanfakossw** [comment]
Great video Daniel!! Looks like a great tool ⭐

**Comment by Hona** [comment]
I hope that Rider adds support for DevTunnels so I can gain that deep IDE integration benefits :) Otherwise, ngrok is pretty easy when I need it. I am planning to use ngrok/devtunnels more for allowing manual testers to test my branch without needing to build or merge and deploy first.

**Comment by PennyWalker** [comment]
That's a really good video Dan! It looks like a cool way to streamline the testing and feedback process for devs

**Comment by sylhuang** [comment]
Very useful tool integrated with VS! Will definitely try this out.

**Comment by AttackOnMorty** [comment]
A good tool for people using VS.

**Comment by SofieHong** [comment]
Love all the tools to simplify the process and improve efficiency! 😆 Also like the 🦋 in the video

**Comment by jerryluossw** [comment]
I think it's really useful, especially in distributed or microservice scenarios,  perform remote debugging.
At the same time, many projects for our government clients are developed by multiple suppliers, so there will be a lot of collaborative remote debugging, I think it will improve our debugging efficiency a lot.

**Comment by Berber-Mama** [comment]
It's the kind of tool that every developer should know about! We're considering translating it and posting it on Bilibili!

**Comment by AnkyChe** [comment]
A useful tool that can save time! The video is really great for a non-developer to understand.

**Comment by johnxu16** [comment]
I cannot wait to use it. The video is very informative and this new feature can speed up the developing processes in so many different scenarios.

**Comment by jimmidier** [comment]
During development of SophieBot the team has experienced scenarios where there was too much of a pain due to difficulty of debugging on Teams client, and that was when port forwarding came in handy. We used ngrok back then and it proved to be saving us plenty of time. Now that the dev tool is released in Visual Studio, I believe it'll be an even better alternative for developers.

**Comment by KristenHu** [comment]
It seems like a useful tool, and the related rule is very clear and detailed, I like it!

**Comment by Freego1783** [comment]
Nice tool to get feedback quickly ! And it seems really easy to set up, I'll probably try it soon.

**Comment by PothieuG** [comment]
This tool looks incredibly practical and easy to set up! I'm discovering the world of port forwarding tools, and the possibilities for use (especially for easily sharing the progress or status of an application) seem essential to me!
The video was clear enough for a novice in the field to understand. Thank you, Daniel !

**Comment by JackDuan123** [comment]
Good Job, Daniel!!!!
 I like the video, it is very handy !!!

**Comment by BrookJeynes** [comment]
Seems like a very valuable tool if you're using VS

**Comment by Marxoz** [comment]
Awesome video Daniel! And awesome feature. Will definitely start playing with it asap!

**Comment by DavidAbraham082** [comment]
Looks very useful, another tool to simplify collaborative work

**Comment by TobyChurches** [comment]
Unreal! I wasn't aware of this feature, but I'll definitely use it now!

**Comment by billyterblanche** [comment]
I am a VS user so quite happy to see this available as it should save time getting a URL ready for someone to use. I don't see any information on the webs (unless my Googling skills are lacking) about any usage limitations? It would be nice to see how it compares to something like ngrok.
